src/main/java/frc/robot/subsystems/drive/CommandSwerveDrivetrain.java

```diff
@@ -173,12 +184,29 @@ private void controlDrivetrain() {
         }
     }
 
-    private static Rotation2d calculateDesiredHeading(Pose2d current, Pose2d target) {
+    private Rotation2d calculateDesiredHeading(Pose2d current, Pose2d target) {
         // I hope this is all inlined
         Pose2d robotToTarget = GeomUtil.transformToPose(current.minus(target));
 
+        double distanceToTarget =
+                Math.hypot(
+                        robotToTarget.getTranslation().getX(),
+                        robotToTarget.getTranslation().getY());
+        Translation2d robotToTargetAnticipated =
+                new Translation2d(
+                        robotToTarget.getX()
+                                + (getRobotVelocity.get().getX()
+                                        * noteTimeToGoal.getValue(distanceToTarget)
+                                        * DriveConstants.velocityScalar),
+                        robotToTarget.getY()
+                                + (getRobotVelocity.get().getY()
+                                        * noteTimeToGoal.getValue(distanceToTarget)
+                                        * DriveConstants.velocityScalar));
+
         Rotation2d angle =
-                Rotation2d.fromRadians(Math.atan2(robotToTarget.getY(), robotToTarget.getX()));
+                Rotation2d.fromRadians(
+                        Math.atan2(
+                                robotToTargetAnticipated.getY(), robotToTargetAnticipated.getX()));
```

Comment:
We need to correct for note velocity here - this is where the distance to target and note time to target is applicable. We have to cancel out the note's lateral velocity by slightly over-turning the robot

Reply:
I added this in the above step instead of here. I can change it to be here though.

Reply:
I interpret the above code as the following:

1. We get the robot's position
2. We predict the robot's position at the time when the note would be hitting the speaker if shot now
3. We compute the goal rotation based on this position

I don't think this works mathematically because then we are likely aiming way too far in the future / at the wrong point in time. I could be wrong though. However, this does motivate us to decide on and describe the sequence of events so we're all on the same page

It would be good to break these calculations into steps with multiple variables just to make it super clear what we are calculating